synthesizer/src/vm/finalize.rs

```diff
@@ -786,6 +790,8 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     /// - The transaction is producing a duplicate output
     /// - The transaction is producing a duplicate transition public key
     /// - The transaction is another deployment in the block from the same public fee payer.
+    /// - The transaction is an execution for a program following its deployment or redeployment in this block.
```

Comment:
I'm still in favour of aborting executions which reference a state root which is older than the latest program upgrade corresponding to that execution.

Reply:
I think this is a fair point. 
The lookup could be done by:
1. Look up the latest transaction ID that deployed or upgraded a program.
2. Look up the corresponding block height for the program.
3. Look up the corresponding block height for the state root.
4. Verify that the state root's block height is less than the program's last deployed/updated height.

One thing to consider is that this check would create a dependence on the state root for "non-record" transactions. It could change the UX for consumer software but is not an unreasonable thing to do.

I'll leave it to reviewers to arbitrate.

Reply:
Sounds great. I can't imagine this to be a major performance bottleneck, but in a next iteration we could consider smarter caching.

CC @raychu86

Reply:
One more reason why this is a good idea:

If a user's local state is behind the network, then they might use an outdated edition or checksum value as witness, and as a result, the proof will fail to verify with an extremely opaque error message.

Now, unfortunately, users are used to transactions failing for opaque reasons. But we can imagine a point where explorers surface this information to users, or developers see this in validator logs for debugging purposes. 

It would be a lot better if the user or developer would see e.g. "Execution failed due to the use of a state root older than the latest program upgrade."

Reply:
For the same reason as my last message above, we should clear executions from the `partially_verified_transactions` cache when the underlying program has been updated.

Reply:
This has been deliberated on a few times. I think it boils down to how we weight usability and user clarity. 

In Ethereum, (to my knowledge) there is no concept of this type of protection when working with upgradable programs via proxy contracts. The conclusion there was to shift the responsibility of safety onto the developer/user in favor of usability.

Reply:
Note that we are already aborting transactions that follow upgrades in the same block. An enforcement of state root "liveness" could cover that case as well.

IMO there is value in providing the safety for users because that is our preference as protocol developers, however it would be valuable to get community and developer feedback as well.

Reply:
I see both arguments for both sides.
As a user, I may want to "know what I am executing". 
On the other hand, a user might feel burdened by the need constantly get the latest state root, especially if the code does not deal with records.

Reply:
My preference is to not add the state root check on transactions in favor of keeping the user experience cleaner and easier.

Particularly for now, unless we see sufficient pushback that favors the more conservative approach.

Reply:
> Perhaps we can pull in ecosystem developers into this code review weigh in?

I like it! I'll share a plan with you.

Note that in any case we will need to clear appropriate executions from `partially_verified_transactions` - though the location where it is cleared might depend on that user feedback.

Reply:
I made a branch with the check ([diff](https://github.com/ProvableHQ/snarkVM/compare/feat/program-updatability..feat/reject-old-execs-after-upgrade))
I'll make a note of it in the tracking issue as well.

Reply:
Refreshed the above branch to do the check in `should_abort_transaction`

Reply:
I genuinely would love to require a recent state root to ensure freshness.

However, I do feel for offline wallets, cold storage schemes, and HSMs (in the future). Namely, that they do not have the privilege or luxury to go and fetch live state like this on a continual basis. And a custodian servicing millions of addresses for their customers does not have the bandwidth to stay on top of live refreshes like this (with today's state of Aleo infrastructure and tooling).

As such, I strongly recommend keeping the design in its current form and re-evaluating down the line.